<a href="https://colab.research.google.com/github/fauziahs13/portfolio_Dec2024/blob/main/Beautiful_Plotly_Charts_%26_Analysing_the_Android_App_Store.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [ ]:
import pandas as pd
import plotly.express as px

# Notebook Presentation

In [ ]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [ ]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [ ]:
df_apps.shape

(10841, 12)

In [ ]:
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Android_Ver'],
      dtype='object')

In [ ]:
df_apps.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0      8.70        0  Paid  $13.99           Teen           Social   
1        0     33.00        0  Paid   $2.99       Everyone        Education   
2        0      5.50        0  Paid   $1.49       Everyone  Personalization   
3        0     19.00        0   NaN       0   Everyone 10+         Strategy   
4        0     10.00        0  Free       0       Everyone         Business   

     Last_Updated         Android_Ver  
0   July 28, 2017          4.1 and up  
1  April 15, 2016          3.0 and up  
2   July 11, 2018          4.2 and up  
3   June 28, 2018  Varies with device  
4  August 6, 2018          4.1 and up

In [ ]:
df_apps.sample(5)

App            Category  Rating  Reviews  \
10240  Need for Speed™ No Limits                GAME    4.40  3344300   
467                     BJ Foods            BUSINESS    5.00        3   
4377        Learn Bulgarian Free              FAMILY    4.70     3049   
1204       Dress Up RagazzA13 DX              FAMILY    3.50       13   
2114                    dz press  NEWS_AND_MAGAZINES     NaN       30   

       Size_MBs    Installs  Type  Price Content_Rating            Genres  \
10240     22.00  50,000,000  Free      0   Everyone 10+            Racing   
467        1.50          10  Free      0       Everyone          Business   
4377      19.00      50,000  Free      0       Everyone         Education   
1204      42.00         100  Paid  $0.99       Everyone        Simulation   
2114      10.00       1,000  Free      0       Everyone  News & Magazines   

             Last_Updated         Android_Ver  
10240       July 24, 2018          4.1 and up  
467      February 7, 2018          4.1 and up  
4377          May 4, 2018  Varies with device  
1204   September 16, 2016        4.0.3 and up  
2114        July 30, 2015          4.0 and up

### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns.

In [ ]:
df_apps.drop(['Last_Updated', 'Android_Ver'], axis=1, inplace=True)
df_apps.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  
0        0      8.70        0  Paid  $13.99           Teen           Social  
1        0     33.00        0  Paid   $2.99       Everyone        Education  
2        0      5.50        0  Paid   $1.49       Everyone  Personalization  
3        0     19.00        0   NaN       0   Everyone 10+         Strategy  
4        0     10.00        0  Free       0       Everyone         Business

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows.

In [ ]:
df_apps.Rating.isna().values.sum()

1474

In [ ]:
# Alternatively
print(df_apps.Rating.isna())
nan_rows = df_apps[df_apps.Rating.isna()]
print(nan_rows.shape)
nan_rows.head()

0         True
1         True
2         True
3         True
4         True
         ...  
10836    False
10837    False
10838    False
10839    False
10840    False
Name: Rating, Length: 10841, dtype: bool
(1474, 10)


App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  
0        0      8.70        0  Paid  $13.99           Teen           Social  
1        0     33.00        0  Paid   $2.99       Everyone        Education  
2        0      5.50        0  Paid   $1.49       Everyone  Personalization  
3        0     19.00        0   NaN       0   Everyone 10+         Strategy  
4        0     10.00        0  Free       0       Everyone         Business

In [ ]:
df_apps_clean = df_apps.dropna()
df_apps_clean.shape

(9367, 10)

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`.


In [ ]:
df_apps_clean.duplicated().values.any()

True

In [ ]:
df_apps_clean.duplicated().values.sum()

476

In [ ]:
# Alternatively
print(df_apps_clean.duplicated())
duplicated_rows = df_apps_clean[df_apps.duplicated()]
print(duplicated_rows.shape)
duplicated_rows.head()

21       False
28       False
47       False
82       False
99       False
         ...  
10836    False
10837    False
10838    False
10839     True
10840    False
Length: 9367, dtype: bool
(476, 10)


<ipython-input-15-9d5b67b5b428>:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  duplicated_rows = df_apps_clean[df_apps.duplicated()]


App Category  Rating  Reviews  \
946                       420 BZ Budeze Delivery  MEDICAL    5.00        2   
1133                                 MouseMingle   DATING    2.70        3   
1196  Cardiac diagnosis (heart rate, arrhythmia)  MEDICAL    4.40        8   
1231                                Sway Medical  MEDICAL    5.00        3   
1247              Chat Kids - Chat Room For Kids   DATING    4.70        6   

      Size_MBs Installs  Type   Price Content_Rating   Genres  
946      11.00      100  Free       0     Mature 17+  Medical  
1133      3.90      100  Free       0     Mature 17+   Dating  
1196      6.50      100  Paid  $12.99       Everyone  Medical  
1231     22.00      100  Free       0       Everyone  Medical  
1247      4.90      100  Free       0     Mature 17+   Dating

In [ ]:
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10808  Instagram   SOCIAL    4.50  66577446      5.30  1,000,000,000  Free   
10809  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10810  Instagram   SOCIAL    4.50  66509917      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10809     0           Teen  Social  
10810     0           Teen  Social

In [ ]:
df_apps_clean = df_apps_clean.drop_duplicates()
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10808  Instagram   SOCIAL    4.50  66577446      5.30  1,000,000,000  Free   
10810  Instagram   SOCIAL    4.50  66509917      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10810     0           Teen  Social

In [ ]:
df_apps_clean = df_apps_clean.drop_duplicates(subset=['App', 'Type', 'Price'])
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social

In [ ]:
df_apps_clean.shape

(8199, 10)

# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [ ]:
df_apps_clean.sort_values('Rating', ascending=False).head()

App     Category  Rating  Reviews  Size_MBs Installs  \
21    KBA-EZ Health Guide      MEDICAL    5.00        4     25.00        1   
1230         Sway Medical      MEDICAL    5.00        3     22.00      100   
1227    AJ Men's Grooming    LIFESTYLE    5.00        2     22.00      100   
1224       FK Dedinje BGD       SPORTS    5.00       36      2.60      100   
1223      CB VIDEO VISION  PHOTOGRAPHY    5.00       13      2.60      100   

      Type Price Content_Rating       Genres  
21    Free     0       Everyone      Medical  
1230  Free     0       Everyone      Medical  
1227  Free     0       Everyone    Lifestyle  
1224  Free     0       Everyone       Sports  
1223  Free     0       Everyone  Photography

INSIGHT --> **Only apps with very few reviews (and a low number on installs)** have perfect 5 star ratings (most likely by friends and family).

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please?

In [ ]:
df_apps_clean.sort_values('Size_MBs', ascending=False).head()

App            Category  Rating  Reviews  \
9942   Talking Babsy Baby: Baby Games           LIFESTYLE    4.00   140995   
10687          Hungry Shark Evolution                GAME    4.50  6074334   
9943            Miami crime simulator                GAME    4.00   254518   
9944     Gangster Town: Vice District              FAMILY    4.30    65146   
3144                       Vi Trainer  HEALTH_AND_FITNESS    3.60      124   

       Size_MBs     Installs  Type Price Content_Rating  \
9942     100.00   10,000,000  Free     0       Everyone   
10687    100.00  100,000,000  Free     0           Teen   
9943     100.00   10,000,000  Free     0     Mature 17+   
9944     100.00   10,000,000  Free     0     Mature 17+   
3144     100.00        5,000  Free     0       Everyone   

                       Genres  
9942   Lifestyle;Pretend Play  
10687                  Arcade  
9943                   Action  
9944               Simulation  
3144         Health & Fitness

INSIGHT --> Here we can clearly see that **there seems to be an upper bound of 100 MB for the size of an app**. A quick google search would also have revealed that **this limit is imposed by the Google Play Store itself**. It’s interesting to see that a number of apps actually hit that limit exactly.

# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [ ]:
highest_reviews = df_apps_clean.sort_values('Reviews', ascending=False)[:50]
highest_reviews.head()

App       Category  Rating  \
10805                                  Facebook         SOCIAL    4.10   
10785                        WhatsApp Messenger  COMMUNICATION    4.40   
10806                                 Instagram         SOCIAL    4.50   
10784  Messenger – Text and Video Chat for Free  COMMUNICATION    4.00   
10650                            Clash of Clans           GAME    4.60   

        Reviews  Size_MBs       Installs  Type Price Content_Rating  \
10805  78158306      5.30  1,000,000,000  Free     0           Teen   
10785  69119316      3.50  1,000,000,000  Free     0       Everyone   
10806  66577313      5.30  1,000,000,000  Free     0           Teen   
10784  56642847      3.50  1,000,000,000  Free     0       Everyone   
10650  44891723     98.00    100,000,000  Free     0   Everyone 10+   

              Genres  
10805         Social  
10785  Communication  
10806         Social  
10784  Communication  
10650       Strategy

In [ ]:
highest_reviews[highest_reviews.Type == 'Paid']

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size_MBs, Installs, Type, Price, Content_Rating, Genres]
Index: []

INSIGHT --> If you look at the number of reviews, you can find the most popular apps on the Android App Store. These include the usual suspects: **Facebook, WhatsApp, Instagram etc**. What’s also notable is that the list of the top 50 most reviewed apps does not include a single paid app! 🤔

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [ ]:
ratings = df_apps_clean.Content_Rating.value_counts()
ratings

Content_Rating
Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: count, dtype: int64

In [ ]:
fig = px.pie(labels=ratings.index, values=ratings.values)
fig.show()

In [ ]:
fig = px.pie(labels=ratings.index, values=ratings.values,
             title="Content Rating",
             names=ratings.index)
fig.update_traces(textposition='outside', textinfo='percent+label')
fig.show()

In [ ]:
fig = px.pie(labels=ratings.index,
values=ratings.values,
title="Content Rating",
names=ratings.index,
hole=0.6,
)
fig.update_traces(textposition='inside', textfont_size=15, textinfo='percent')

fig.show()

# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install?

Check the datatype of the Installs column.

Count the number of apps at each level of installations.

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first.

In [ ]:
# df_apps_clean.sort_values('Installs', ascending=False) #THIS is not recommeded, cause the highest was 500 million, not 1 billion
print(df_apps_clean[df_apps_clean.Installs == "1"].count())
df_apps_clean[df_apps_clean.Installs == "1,000,000,000"].count()

App               3
Category          3
Rating            3
Reviews           3
Size_MBs          3
Installs          3
Type              3
Price             3
Content_Rating    3
Genres            3
dtype: int64


App               20
Category          20
Rating            20
Reviews           20
Size_MBs          20
Installs          20
Type              20
Price             20
Content_Rating    20
Genres            20
dtype: int64

In [ ]:
print(type(df_apps_clean.Installs[21]))
df_apps_clean.Installs.describe()

<class 'str'>


count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

In [ ]:
df_apps_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   object 
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   object 
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 962.6+ KB


In [ ]:
df_apps_clean.groupby('Installs').count()

App  Category  Rating  Reviews  Size_MBs  Type  Price  \
Installs                                                                
1                 3         3       3        3         3     3      3   
1,000           698       698     698      698       698   698    698   
1,000,000      1417      1417    1417     1417      1417  1417   1417   
1,000,000,000    20        20      20       20        20    20     20   
10               69        69      69       69        69    69     69   
10,000          988       988     988      988       988   988    988   
10,000,000      933       933     933      933       933   933    933   
100             303       303     303      303       303   303    303   
100,000        1096      1096    1096     1096      1096  1096   1096   
100,000,000     189       189     189      189       189   189    189   
5                 9         9       9        9         9     9      9   
5,000           425       425     425      425       425   425    425   
5,000,000       607       607     607      607       607   607    607   
50               56        56      56       56        56    56     56   
50,000          457       457     457      457       457   457    457   
50,000,000      202       202     202      202       202   202    202   
500             199       199     199      199       199   199    199   
500,000         504       504     504      504       504   504    504   
500,000,000      24        24      24       24        24    24     24   

               Content_Rating  Genres  
Installs                               
1                           3       3  
1,000                     698     698  
1,000,000                1417    1417  
1,000,000,000              20      20  
10                         69      69  
10,000                    988     988  
10,000,000                933     933  
100                       303     303  
100,000                  1096    1096  
100,000,000               189     189  
5                           9       9  
5,000                     425     425  
5,000,000                 607     607  
50                         56      56  
50,000                    457     457  
50,000,000                202     202  
500                       199     199  
500,000                   504     504  
500,000,000                24      24

In [ ]:
df_apps_clean.Installs = df_apps_clean.Installs.astype(str).str.replace(',', "")
df_apps_clean.Installs = pd.to_numeric(df_apps_clean['Installs'])
print(type(df_apps_clean.Installs[21]))
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

<class 'numpy.int64'>


App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [ ]:
df_apps_clean.Price.describe()

count     8199
unique      73
top          0
freq      7595
Name: Price, dtype: object

In [ ]:
df_apps_clean.Price = df_apps_clean.Price.astype(str).str.replace('$', "")
df_apps_clean.Price = pd.to_numeric(df_apps_clean['Price'])
print(type(df_apps_clean.Price[21]))
# df_apps_clean[['App', 'Price']].groupby('Price').count()

<class 'numpy.float64'>


### The most expensive apps sub $250

In [ ]:
df_apps_clean = df_apps_clean[df_apps_clean['Price'] < 250]
df_apps_clean.sort_values('Price', ascending=False).head(5)

App   Category  Rating  Reviews  Size_MBs  \
2281  Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407               LTC AS Legal    MEDICAL    4.00        6      1.30   
2629           I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481    A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
2463         PTA Content Master    MEDICAL    4.20       64     41.00   

      Installs  Type  Price Content_Rating     Genres  
2281      1000  Paid  79.99       Everyone    Medical  
1407       100  Paid  39.99       Everyone    Medical  
2629      1000  Paid  37.99       Everyone  Lifestyle  
2481      1000  Paid  33.99       Everyone    Medical  
2463      1000  Paid  29.99       Everyone    Medical

### Highest Grossing Paid Apps (ballpark estimate)

In [ ]:
df_apps_clean['Revenue_Estimate'] = df_apps_clean.Installs.mul(df_apps_clean.Price)
df_apps_clean.sort_values('Revenue_Estimate', ascending=False)[:10]

App     Category  Rating  Reviews  Size_MBs  \
9220                      Minecraft       FAMILY    4.50  2376564     19.00   
8825                  Hitman Sniper         GAME    4.60   408292     29.00   
7151  Grand Theft Auto: San Andreas         GAME    4.40   348962     26.00   
7477            Facetune - For Free  PHOTOGRAPHY    4.40    49553     48.00   
7977        Sleep as Android Unlock    LIFESTYLE    4.50    23966      0.85   
6594            DraStic DS Emulator         GAME    4.60    87766     12.00   
6082                   Weather Live      WEATHER    4.50    76593      4.75   
7954                    Bloons TD 5       FAMILY    4.60   190086     94.00   
7633        Five Nights at Freddy's         GAME    4.60   100805     50.00   
6746     Card Wars - Adventure Time       FAMILY    4.30   129603     23.00   

      Installs  Type  Price Content_Rating                     Genres  \
9220  10000000  Paid   6.99   Everyone 10+  Arcade;Action & Adventure   
8825  10000000  Paid   0.99     Mature 17+                     Action   
7151   1000000  Paid   6.99     Mature 17+                     Action   
7477   1000000  Paid   5.99       Everyone                Photography   
7977   1000000  Paid   5.99       Everyone                  Lifestyle   
6594   1000000  Paid   4.99       Everyone                     Action   
6082    500000  Paid   5.99       Everyone                    Weather   
7954   1000000  Paid   2.99       Everyone                   Strategy   
7633   1000000  Paid   2.99           Teen                     Action   
6746   1000000  Paid   2.99   Everyone 10+    Card;Action & Adventure   

      Revenue_Estimate  
9220     69,900,000.00  
8825      9,900,000.00  
7151      6,990,000.00  
7477      5,990,000.00  
7977      5,990,000.00  
6594      4,990,000.00  
6082      2,995,000.00  
7954      2,990,000.00  
7633      2,990,000.00  
6746      2,990,000.00

INSIGHT --> The top spot of the highest-grossing paid app goes to … **Minecraft at close to $70 million**. It’s quite interesting that Minecraft (along with Bloons and Card Wars) is actually listed in the Family category rather than in the Game category. If we include these titles, **we see that 7 out the top 10 highest-grossing apps are games.** The Google Play Store seems to be quite flexible with its category labels.

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

If you were to release an app, you might want to target a category which is both popular but also one where the downloads are spread out among many different apps.

In [ ]:
top10_category = df_apps_clean.Category.value_counts()[:10]
top10_category

Category
FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: count, dtype: int64

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [ ]:
bar = px.bar(x = top10_category.index, # index = category name
             y = top10_category.values)

bar.show()

INSIGHT --> Based on the number of apps, **the Family and Game categories are the most competitive.** Releasing yet another app into these categories will make it hard to get noticed.

But what if we look at it from a different perspective? What matters is not just the total number of apps in the category but how often apps are downloaded in that category. This will give us an idea of how popular a category is. First, we have to group all our apps by category and sum the number of installations:

In [ ]:
category_installs = df_apps_clean.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace=True)

In [ ]:
category_installs

Installs
Category                        
EVENTS                  15949410
BEAUTY                  26916200
PARENTING               31116110
MEDICAL                 39162676
COMICS                  44931100
LIBRARIES_AND_DEMO      52083000
AUTO_AND_VEHICLES       53129800
HOUSE_AND_HOME          97082000
ART_AND_DESIGN         114233100
DATING                 140912410
FOOD_AND_DRINK         211677750
EDUCATION              352852000
WEATHER                361096500
FINANCE                455249400
MAPS_AND_NAVIGATION    503267560
LIFESTYLE              503611120
BUSINESS               692018120
SPORTS                1096431465
HEALTH_AND_FITNESS    1134006220
SHOPPING              1400331540
PERSONALIZATION       1532352930
BOOKS_AND_REFERENCE   1665791655
ENTERTAINMENT         2113660000
NEWS_AND_MAGAZINES    2369110650
TRAVEL_AND_LOCAL      2894859300
VIDEO_PLAYERS         3916897200
FAMILY                4437554490
PHOTOGRAPHY           4649143130
SOCIAL                5487841475
PRODUCTIVITY          5788070180
TOOLS                 8099724500
COMMUNICATION        11039241530
GAME                 13858762717

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [ ]:
h_bar = px.bar(x = category_installs.Installs,
               y = category_installs.index,
               orientation='h',
               title='Category Popularity')

h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')

h_bar.show()

INSIGHT --> Now we see that **Games and Tools (and Communication) are actually the most popular categories**. If we plot the popularity of a category next to the number of apps in that category we can get an idea of how concentrated a category is.

Do few apps have most of the downloads or are the downloads spread out over many apps?

### Category Concentration - Downloads vs. Competition

**Challenge**:
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:

<img src=https://imgur.com/uQRSlXi.png width="350">

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this.

<img src=https://imgur.com/cHsqh6a.png>

*Hint*: Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log')

In [ ]:
cat_number = df_apps_clean.groupby('Category').agg({'App': pd.Series.count})
cat_merged_df = pd.merge(cat_number, category_installs, on='Category', how="inner")
print(f'The dimensions of the DataFrame are: {cat_merged_df.shape}')
cat_merged_df.sort_values('Installs', ascending=False)

The dimensions of the DataFrame are: (33, 2)


App     Installs
Category                              
GAME                  910  13858762717
COMMUNICATION         257  11039241530
TOOLS                 719   8099724500
PRODUCTIVITY          301   5788070180
SOCIAL                203   5487841475
PHOTOGRAPHY           263   4649143130
FAMILY               1606   4437554490
VIDEO_PLAYERS         148   3916897200
TRAVEL_AND_LOCAL      187   2894859300
NEWS_AND_MAGAZINES    204   2369110650
ENTERTAINMENT         102   2113660000
BOOKS_AND_REFERENCE   169   1665791655
PERSONALIZATION       298   1532352930
SHOPPING              180   1400331540
HEALTH_AND_FITNESS    243   1134006220
SPORTS                260   1096431465
BUSINESS              262    692018120
LIFESTYLE             297    503611120
MAPS_AND_NAVIGATION   118    503267560
FINANCE               296    455249400
WEATHER                72    361096500
EDUCATION             118    352852000
FOOD_AND_DRINK         94    211677750
DATING                134    140912410
ART_AND_DESIGN         61    114233100
HOUSE_AND_HOME         62     97082000
AUTO_AND_VEHICLES      73     53129800
LIBRARIES_AND_DEMO     64     52083000
COMICS                 54     44931100
MEDICAL               292     39162676
PARENTING              50     31116110
BEAUTY                 42     26916200
EVENTS                 45     15949410

In [ ]:
# Alternatively
# category_installs['App'] = df_apps_clean.groupby('Category').agg({'App': pd.Series.count})

In [ ]:
scatter = px.scatter(cat_merged_df, # data
                    x='App', # column name
                    y='Installs',
                    title='Category Concentration',
                    size='App',
                    hover_name=cat_merged_df.index, color='Installs')

scatter.update_layout(xaxis_title="Number of Apps (Lower=More Concentrated)",
                      yaxis_title="Installs",
                      yaxis=dict(type='log'))

scatter.show()

INSIGHT ==> What we see is that the categories like Family, Tools, and Game have many different apps sharing a high number of downloads which are **Diverse & Popular (Top Right)**. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps which are **Concentrated & Popular (Top Left)**. And Bottom Left is apps that Concentrated & Not Popular/ Unloved lol.

# Extracting Nested Data from a Column

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html).


In [ ]:
# df_apps_clean[['App', 'Genres']].groupby('Genres').value_counts()
print(len(df_apps_clean.Genres.unique()))

114


In [ ]:
df_apps_clean.Genres.value_counts()

Genres
Tools                                718
Entertainment                        467
Education                            429
Productivity                         301
Personalization                      298
                                    ... 
Adventure;Brain Games                  1
Travel & Local;Action & Adventure      1
Art & Design;Pretend Play              1
Music & Audio;Music & Video            1
Lifestyle;Pretend Play                 1
Name: count, Length: 114, dtype: int64

In [ ]:
# Split the strings on the semi-colon and then .stack them.
stack = df_apps_clean.Genres.str.split(';', expand=True).stack()
print(f'We now have a single column with shape: {stack.shape}')
stack

We now have a single column with shape: (8564,)


21     0                    Medical
28     0                     Arcade
47     0                     Arcade
82     0                     Arcade
99     0                    Medical
                     ...           
10824  0               Productivity
10828  0    Video Players & Editors
10829  0    Video Players & Editors
10831  0           News & Magazines
10835  0                     Arcade
Length: 8564, dtype: object

In [ ]:
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')
# print(f'Number of genres: {num_genres.shape}')
num_genres

Number of genres: 53


Tools                      719
Education                  587
Entertainment              498
Action                     304
Productivity               301
Personalization            298
Lifestyle                  298
Finance                    296
Medical                    292
Sports                     270
Photography                263
Business                   262
Communication              258
Health & Fitness           245
Casual                     216
News & Magazines           204
Social                     203
Simulation                 200
Travel & Local             187
Arcade                     185
Shopping                   180
Books & Reference          171
Video Players & Editors    150
Dating                     134
Puzzle                     124
Maps & Navigation          118
Role Playing               111
Racing                     103
Action & Adventure          96
Strategy                    95
Food & Drink                94
Educational                 93
Adventure                   78
Auto & Vehicles             73
Weather                     72
Pretend Play                68
Brain Games                 65
Libraries & Demo            64
Art & Design                62
House & Home                62
Board                       57
Comics                      54
Parenting                   50
Card                        46
Events                      45
Beauty                      42
Casino                      37
Music & Video               31
Creativity                  31
Trivia                      28
Word                        22
Music                       21
Music & Audio                1
Name: count, dtype: int64

In [ ]:
# num_genres.columns

# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create this chart with the Series containing the genre data?

<img src=https://imgur.com/DbcoQli.png width=400>

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/).

* Find a way to set the colour scale using the color_continuous_scale parameter.
* Find a way to make the color axis disappear by using coloraxis_showscale.

In [ ]:
genre_bar = px.bar(x = num_genres.index[:15], # index = genre name
                  y = num_genres.values[:15], # count
                  title='Top Genres',
                  hover_name=num_genres.index[:15],
                  color=num_genres.values[:15],
                  color_continuous_scale='Agsunset')

genre_bar.update_layout(xaxis_title='Genre', yaxis_title='Number of Apps', coloraxis_showscale=False)

genre_bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

In [ ]:
df_apps_clean.Type.value_counts()

Type
Free    7595
Paid     589
Name: count, dtype: int64

In [ ]:
df_free_vs_paid = df_apps_clean.groupby(["Category", "Type"], as_index=False).agg({'App': pd.Series.count})
df_free_vs_paid.sort_values('App')

Category  Type   App
3    AUTO_AND_VEHICLES  Paid     1
24      FOOD_AND_DRINK  Paid     2
38  NEWS_AND_MAGAZINES  Paid     2
40           PARENTING  Paid     2
17       ENTERTAINMENT  Paid     2
..                 ...   ...   ...
31           LIFESTYLE  Free   284
21             FINANCE  Free   289
53               TOOLS  Free   656
25                GAME  Free   834
19              FAMILY  Free  1456

[61 rows x 3 columns]

In [ ]:
df_free_vs_paid[df_free_vs_paid.Category == 'HEALTH_AND_FITNESS']

Category  Type  App
27  HEALTH_AND_FITNESS  Free  232
28  HEALTH_AND_FITNESS  Paid   11

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create this bar chart:

<img src=https://imgur.com/LE0XCxA.png>

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category.

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value).

In [ ]:
g_bar = px.bar(df_free_vs_paid, x='Category',
               y='App',
               title='Free vs Paid Apps by Category',
               color='Type',
               barmode='group',)
# Change the bar mode
g_bar.update_layout(xaxis_title='Category', yaxis_title='Number of Apps',
                    xaxis= {'categoryorder':'total descending'},
                    yaxis=dict(type='log'))

g_bar.show()

INSIGHT --> What we see is that while there are very few paid apps on the Google Play Store, **some categories have relatively more paid apps than others, including Personalization, Medical and Weather**. So, depending on the category you are targeting, it might make sense to release a paid-for app.

# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the following chart.

<img src=https://imgur.com/uVsECT3.png>


In [ ]:
# df_free_vs_paid_ins = df_apps_clean.groupby(["Category", "Type"], as_index=False).agg({'Installs': pd.Series.sum})
# df_free_vs_paid_ins.sort_values('Installs')
df_apps_clean.groupby('Type').agg({'Installs': pd.Series.sum})

Installs
Type             
Free  75121934856
Paid     57132382

In [ ]:
box = px.box(df_apps_clean, x='Type',
             y='Installs',
             title='How Many Downloads are Paid Apps Giving Up?',
             color='Type',
             points="all",
             notched=True, # used notched shape
            )

box.update_layout(xaxis_title='Type', yaxis_title='Installs',
                  yaxis=dict(type='log'))

box.show()

INSIGHT --> From the hover text in the chart, we see that the median number of downloads for free apps is **500,000**, while the median number of downloads for paid apps is around **5,000**! **This is massively lower.** But does this mean we should give up on selling a paid app? Let’s see how much revenue we would estimate per category.



# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate the chart below:

<img src=https://imgur.com/v4CiNqX.png>

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: I've used 'min ascending' to sort the categories.

In [ ]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
revenue_box = px.box(df_paid_apps, x='Category',
                     y='Revenue_Estimate',
                     title='How Much Can Paid App Earn?'
                     )

revenue_box.update_layout(xaxis_title='Category',
                          yaxis_title='Paid App Ballpark Revenue',
                          xaxis={'categoryorder':'min ascending'},
                          yaxis=dict(type='log'))

revenue_box.show()

INSIGHT --> If an Android app costs USD30,000 to develop, then the average app in **very few categories would cover that development cost**. The median paid photography app earned about $20,000. **Many more app’s revenues were even lower (earned below USD10k)** - meaning they would need other sources of revenue like advertising or in-app purchases to make up for their development costs. However, certain app categories seem to contain a large number of **outliers** that have much higher (estimated) revenue - for example in **Medical, Personalisation, Tools, Game, and Family**.



# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [ ]:
df_paid_apps.Price.median()

2.99

In [ ]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
price_box = px.box(df_paid_apps, x='Category',
                     y='Price',
                     title='Price Per Category'
                     )

price_box.update_layout(xaxis_title='Category',
                          yaxis_title='Paid App Price',
                          xaxis={'categoryorder':'max descending'},
                          yaxis=dict(type='log'))

price_box.show()

INSIGHT --> However, some categories have higher median prices than others. This time we see that **Medical apps have the most expensive apps as well as a median price of USD5.49**. In contrast, Personalisation apps are quite cheap on average at USD1.49. Other categories which higher median prices are Business (USD4.99) and Dating ($6.99). It seems like customers who shop in these categories are not so concerned about paying a bit extra for their apps.

INSIGHT --> If the average overall is USD 2.99, Categories with High average price would be Medical, Business, Dating, (maybe also Finance, Weather, Education, Parenting, Food_and_Drink, Entertaiment)